Allow for retry when tasks are stuck in queued

dimberman opened the pull request:
Tasks can get stuck in queued for a wide variety of reasons (e.g. celery loses
track of a task, a cluster can't further scale up its workers, etc.), but tasks
should not be stuck in queued for a long time.

Originally, we simply marked a task as failed when it was stuck in queued for
too long. We found that this led to suboptimal outcomes as ideally we would like "failed"
to mean that a task was unable to run, instead of it meaning that we were unable to run the task.

As a compromise between always failing a stuck task and always rescheduling a stuck task (which could
lead to tasks being stuck in queued forever without informing the user), we have creating the config
`[core] num_stuck_reschedules`. With this new configuration, an airflow admin can decide how
sensitive they would like their airflow to be WRT failing stuck tasks.


Here is an example of what it looks like after trying this out with celery executor

<img width="1109" alt="image" src="https://github.com/user-attachments/assets/233af08e-703a-4b1f-82e1-3340c52bf43a">

dstandish commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1822,9 +1822,30 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
+    def _reset_task_instance(self, ti: TaskInstance, session: Session = NEW_SESSION):
+        ti.external_executor_id = None
+        ti.state = State.SCHEDULED
+        session.add(ti)
```
Merge not add

o-nikolas commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1822,9 +1822,30 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
                                 ),
                             )
                         )
```
Can you update the wording in the logging and docstring above, given the new behaviour.

dimberman replied:
Yes I've created separate logs for rescheduling and failing of tasks

dstandish commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1822,9 +1822,33 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
+                        .where(Log.run_id == ti.run_id)
+                        .where(Log.map_index == ti.map_index)
+                        .where(Log.try_number == ti.try_number)
+                        .where(Log.event == "stuck in queued")
```
should be constant of some kind to ensure that the events are created / looked up with same text

dimberman replied:
Agreed, was on my list to do before taking PR out of draft :)

dstandish commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1822,9 +1822,33 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
+                        .where(Log.run_id == ti.run_id)
+                        .where(Log.map_index == ti.map_index)
+                        .where(Log.try_number == ti.try_number)
+                        .where(Log.event == "stuck in queued")
```
simpler / cleaner i think to just use the `*args` e.g.
```suggestion
.where(
    Log.dag_id == ti.dag_id,
    Log.run_id == ti.run_id,
    Log.map_index == ti.map_index,
    Log.try_number == ti.try_number,
    Log.event == "stuck in queued",
)
```

dimberman committed:
Handle the scenario where a task is queued for longer than `task_queued_timeout`.

Tasks can get stuck in queued for a wide variety of reasons (e.g. celery loses
track of a task, a cluster can't further scale up its workers, etc.), but tasks
should not be stuck in queued for a long time.

Originally, we simply marked a task as failed when it was stuck in queued for
too long. We found that this led to suboptimal outcomes as ideally we would like "failed"
to mean that a task was unable to run, instead of it meaning that we were unable to run the task.

As a compromise between always failing a stuck task and always rescheduling a stuck task (which could
lead to tasks being stuck in queued forever without informing the user), we have creating the config
`AIRFLOW__CORE__NUM_STUCK_RETRIES`. With this new configuration, an airflow admin can decide how
sensitive they would like their airflow to be WRT failing stuck tasks.

dimberman committed:
address feedback

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1817,17 +1828,59 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
                                 task_instance=ti.key,
                                 extra=(
-                                    "Task will be marked as failed. If the task instance has "
+                                    f"Task was stuck in queued and will be retried, once it has hit {num_allowed_retries} attempts"
```
```suggestion
                                    f"Task was stuck in queued and will be requeued, once it has hit {num_allowed_retries} attempts"
```

It might be good to avoid "retry" here to avoid confusion between dag-author-configured-retries and these retries.

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1805,6 +1815,7 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
             )
         ).all()
 
+        num_allowed_retries = conf.getint("core", "num_stuck_retries", fallback=2)
```
```suggestion
        num_allowed_retries = conf.getint("core", "num_stuck_retries")
```

You need to add this to [config.yml](https://github.com/apache/airflow/blob/main/airflow/config_templates/config.yml) (and then won't need a fallback).

jedcunningham commented on providers/src/airflow/providers/celery/executors/celery_executor.py:
```diff
@@ -450,7 +450,6 @@ def cleanup_stuck_queued_tasks(self, tis: list[TaskInstance]) -> list[str]:
         for ti in tis:
             readable_tis.append(repr(ti))
             task_instance_key = ti.key
-            self.fail(task_instance_key, None)
```
This is problematic - folks could install the provider with this code onto 2.8.

dimberman replied:
@jedcunningham can you explain a bit further? Is there a workaround we can consider? The reason we did this is because the idea of failing the task and then setting it to scheduled after felt fraught with potential bugs

dstandish replied:
it's possible in the provider to inspect the airflow version and handle conditionally

dimberman replied:
@dstandish yep added exactly that.

jedcunningham commented on uv.lock:
I assume you didn't intend to add this :)

dimberman replied:
Oh weird, I explicitly deleted this and commited, I'll add that to my global gitignore

dimberman committed:
remove uv.lock

dimberman committed:
Update airflow/jobs/scheduler_job_runner.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dimberman committed:
We need to ensure that older versions of airflow don't run into issues since they will not have the scheduler change

dimberman committed:
Merge branch 'handle-stuck-in-queue' of https://github.com/apache/airflow into handle-stuck-in-queue

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1785,15 +1787,23 @@ def _send_dag_callbacks_to_processor(self, dag: DAG, callback: DagCallbackReques
+
+        As a compromise between always failing a stuck task and always rescheduling a stuck task (which could
+        lead to tasks being stuck in queued forever without informing the user), we have creating the config
+        `AIRFLOW__CORE__NUM_STUCK_RETRIES`. With this new configuration, an airflow admin can decide how
```
```suggestion
        `[core] num_stuck_reschedules`. With this new configuration, an airflow admin can decide how
```

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -98,6 +98,8 @@
 DR = DagRun
 DM = DagModel
 
+RETRY_STUCK_IN_QUEUED_EVENT = "retrying stuck in queued"
```
```suggestion
RESCHEDULE_STUCK_IN_QUEUED_EVENT = "rescheduling stuck in queued"
```

etc.

dimberman committed:
Merge branch 'main' into handle-stuck-in-queue

dimberman committed:
address feedback

dimberman committed:
Merge branch 'handle-stuck-in-queue' of https://github.com/apache/airflow into handle-stuck-in-queue

dimberman committed:
pre-commit

dimberman committed:
Update airflow/jobs/scheduler_job_runner.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dimberman committed:
Merge branch 'main' into handle-stuck-in-queue

jedcunningham commented on airflow/config_templates/config.yml:
```diff
@@ -386,6 +386,14 @@ core:
+      description: |
+        Number of times Airflow will reschedule a task that gets stuck in queued before marking the
+        task as failed
+      version_added: 2.10.0
```
```suggestion
      version_added: 2.10.4
```

dimberman committed:
Handle the scenario where a task is queued for longer than `task_queued_timeout`.

Tasks can get stuck in queued for a wide variety of reasons (e.g. celery loses
track of a task, a cluster can't further scale up its workers, etc.), but tasks
should not be stuck in queued for a long time.

Originally, we simply marked a task as failed when it was stuck in queued for
too long. We found that this led to suboptimal outcomes as ideally we would like "failed"
to mean that a task was unable to run, instead of it meaning that we were unable to run the task.

As a compromise between always failing a stuck task and always rescheduling a stuck task (which could
lead to tasks being stuck in queued forever without informing the user), we have creating the config
`AIRFLOW__CORE__NUM_STUCK_RETRIES`. With this new configuration, an airflow admin can decide how
sensitive they would like their airflow to be WRT failing stuck tasks.

dimberman committed:
Merge branch 'handle-stuck-in-queue' of https://github.com/apache/airflow into handle-stuck-in-queue

dimberman committed:
Update airflow/config_templates/config.yml

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dimberman committed:
k8s support

dimberman committed:
Merge branch 'handle-stuck-in-queue' of https://github.com/apache/airflow into handle-stuck-in-queue

dimberman committed:
Merge branch 'main' of https://github.com/apache/airflow into handle-stuck-in-queue

dstandish commented on providers/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -636,6 +638,10 @@ def cleanup_stuck_queued_tasks(self, tis: list[TaskInstance]) -> list[str]:
                 self.log.warning("Cannot find pod for ti %s", ti)
                 continue
             readable_tis.append(repr(ti))
+            if Version(airflow_version) < Version("2.10.4"):
```
should this be a greater than?

dimberman committed:
Merge branch 'main' of https://github.com/apache/airflow into handle-stuck-in-queue

dimberman committed:
tested against kubernetes_executor

dstandish committed:
Simplify the handle stuck in queued interface (#43647)

dimberman committed:
fix

dimberman committed:
Merge branch 'handle-stuck-in-queue' of https://github.com/apache/airflow into handle-stuck-in-queue

dimberman committed:
Merge branch 'main' of https://github.com/apache/airflow into handle-stuck-in-queue

dstandish committed:
remove commented code

dstandish committed:
simplify / make more acurrate the log info in backcompat case

dstandish committed:
remove redundant event log

dimberman committed:
cleanup

dimberman committed:
nit

dstandish committed:
fix test

dstandish committed:
fix test

dstandish committed:
handle when func not implemented

dstandish committed:
reorder the log add to follow the actual requeue

dstandish committed:
remove obsolete event

dstandish committed:
tweak event language again

dstandish committed:
just remove from running set instead of pretending that we are changing state

dimberman committed:
Merge branch 'main' of https://github.com/apache/airflow into handle-stuck-in-queue

dimberman committed:
Merge branch 'handle-stuck-in-queue' of https://github.com/apache/airflow into handle-stuck-in-queue

dimberman committed:
fix test

dimberman committed:
fix test

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1796,28 +1808,95 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
+                    )
+                    continue
+
+                self.log.warning("Task stuck in queued and may be requeued. task_id=%s", ti.key)
```
```suggestion
                self.log.warning("Task stuck in queued and may be requeued. ti_key=%s", ti.key)
```

nit

dstandish replied:
updated

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1796,28 +1808,95 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
+                    )
+                    with suppress(KeyError):
+                        executor.running.remove(ti.key)
+                    self._reschedule_stuck_task(ti)
```
Do we also want a logging line for this (vs just a new `Log` record)?

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1796,28 +1808,95 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
+        )
+
+    @provide_session
+    def _reset_task_instance(self, ti: TaskInstance, session: Session = NEW_SESSION):
```
this isn't used anymore?

dstandish replied:
fixed

jedcunningham commented on providers/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py:
```diff
@@ -443,6 +446,23 @@ def delete_pod(self, pod_name: str, namespace: str) -> None:
                 raise
 
+    def patch_pod_delete_stuck(self, *, pod_name: str, namespace: str):
+        """Add a "done" annotation to ensure we don't continually adopt pods."""
```
docstring was never updated after copy/paste

dstandish replied:
fixed

jedcunningham commented on airflow/executors/base_executor.py:
```diff
@@ -540,7 +540,9 @@ def terminate(self):
-    def cleanup_stuck_queued_tasks(self, tis: list[TaskInstance]) -> list[str]:  # pragma: no cover
+    def cleanup_stuck_queued_tasks(
+        self, tis: list[TaskInstance]
+    ) -> Iterable[TaskInstance]:  # pragma: no cover
```
I'm a bit concerned about this return type change (str -> TI), and also the behavior change for this method. BaseExecutor is public, so we have to consider executors beyond those in our repos. And today, we expect other executors to fail tasks in this method.

dstandish replied:
outdated

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1796,28 +1808,95 @@ def _fail_tasks_stuck_in_queued(self, session: Session = NEW_SESSION) -> None:
+                    #   this is for backcompat. the pre-2.10.4 version of the interface
+                    #   expected a string return val.
+                    self.log.warning(
+                        "Task instance %s stuck in queued.  May be set to failed.",
```
If we get a str back, shouldn't we assume it was set to failed? That's the current expectation, right?

dstandish replied:
outdated; we added new func

jedcunningham commented on airflow/executors/base_executor.py:
```diff
@@ -540,7 +540,9 @@ def terminate(self):
+        self, tis: list[TaskInstance]
+    ) -> Iterable[TaskInstance]:  # pragma: no cover
         """
         Handle remnants of tasks that were failed because they were stuck in queued.
```
We should reword this docsrting too.

dstandish replied:
no longer changing this function's behavior

dstandish committed:
Merge branch 'main' into handle-stuck-in-queue

dstandish committed:
Revert signature change in `cleanup_stuck_queued_tasks`

dstandish committed:
Revert "Revert signature change in `cleanup_stuck_queued_tasks`"

This reverts commit f475d124bb9f514c7b1b44a74339fe8678e32032.

dstandish committed:
Refactor this beast to deprecate old function and add new one

dstandish committed:
fix tests

dstandish committed:
cleanups

dstandish committed:
fix readability

dstandish committed:
mark as deprecated / add todo

dstandish committed:
move log line

dstandish committed:
fix test

dstandish committed:
docstring / spelling

dstandish committed:
fix test

dstandish committed:
Merge branch 'main' into handle-stuck-in-queue

dstandish committed:
fix tests

dstandish committed:
don't create an executor event; just fail the task

eladkal commented on airflow/config_templates/config.yml:
```diff
@@ -2331,6 +2331,14 @@ scheduler:
       type: integer
       example: ~
       default: "8974"
+    num_stuck_in_queued_retries:
```
I am late to join the party but I don't follow why we need this. We have [AIRFLOW__SCHEDULER__TASK_QUEUED_TIMEOUT](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#task-queued-timeout) introduced in Airflow 2.6
Which means that task retries after timeout passed. If for some reason this setting isn't doing it now the we should fix it rather than adding a new setting.

`task_queued_timeout` + `num_stuck_in_queued_retries` feels too complicated.

collinmcnulty replied:
task queued timeout consumes the retries on the task as set by the dag author. This is a problem when the task is not idempotent and therefore deliberately has retries set to 0. But the particular failure of a task to even get to running should not actually consume the task's retries that are for the running of the task. So we still need the timeout setting, but we also need a separate counter for attempts to get from queued to running, whereas retries are attempts to get from running to success.

dstandish replied:
so yeah, these are very much different things.
do we need `num_stuck_in_queued_retries` to be configurable? not necessarily.

eladkal replied:
> so yeah, these are very much different things.
do we need num_stuck_in_queued_retries to be configurable? not necessarily.

That is where I was going. This sounds more like airflow internal mechanism. I think we better not expose this and possibly cause confusion how settings are interacting and what they are about

dstandish replied:
Alright.  I have made this an undocumented param.  Hopefully this pleases everyone.  If a sophisticated user (like an airflow-as-a-service provider) wanted to customize this, they can.  But for average user, it is invisible.

And making it configurable-albeit-hidden is a bonus for test readability because we can control exactly how many time it should retry, and this is important to control for testing.  Resolving convo.  PTAL

dstandish committed:
Merge branch 'main' into handle-stuck-in-queue

dstandish committed:
Change this from "clean up stuck queued" to "revoke_task"

dstandish committed:
update tests

dstandish committed:
fix docstring

dstandish committed:
docstring

dstandish committed:
make the param undocumented

dstandish committed:
fix fallback

dstandish committed:
ensure task removed from running / queued

dstandish committed:
update test

dstandish committed:
fix test mistake

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1767,48 +1780,137 @@ def _send_dag_callbacks_to_processor(self, dag: DAG, callback: DagCallbackReques
+        too long. We found that this led to suboptimal outcomes as ideally we would like "failed"
+        to mean that a task was unable to run, instead of it meaning that we were unable to run the task.
+
+        As a compromise between always failing a stuck task and always rescheduling a stuck task (which could
```
We don't way what the compromise is. Should reword this a bit.

dstandish replied:
yea

jedcunningham commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1767,48 +1780,137 @@ def _send_dag_callbacks_to_processor(self, dag: DAG, callback: DagCallbackReques
+        with suppress(NotImplementedError):
+            for ti_repr in executor.cleanup_stuck_queued_tasks(tis=stuck_tis):
+                self.log.warning(
+                    "Task instance %s stuck in queued.  May be set to failed.",
```
If we are calling `cleanup_stuck_queued_tasks`, won't it be set to failed, not "may be"?

dstandish replied:
yea

jedcunningham commented on providers/src/airflow/providers/celery/executors/celery_executor.py:
```diff
@@ -433,31 +434,34 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
         return not_adopted_tis
 
+    @deprecated(
+        reason="Replaced by function `revoke_task`. Upgrade airflow core to make this go away.",
```
hmm, I wonder if we should instead not deprecate this, but leave a comment that it can be removed when the min airflow version for providers is 2.11?

dstandish replied:
Why not deprecate?  This warning would only appear if user upgrades celery but not airflow

jedcunningham replied:
Yeah. It doesn't really matter. Just feels a bit cleaner to me.

We should at least state what version to upgrade to.

dstandish replied:
we don't know what version this is gonna be released in, alas

dstandish replied:
it would also only show up in scheduler logs, and only when stuck quueued tasks processed

dstandish committed:
small nits

jscheffl commented on airflow/executors/base_executor.py:
```diff
@@ -552,7 +554,12 @@ def terminate(self):
         raise NotImplementedError
 
-    def cleanup_stuck_queued_tasks(self, tis: list[TaskInstance]) -> list[str]:  # pragma: no cover
+    @deprecated(
```
@dstandish 

As this is a change of (public) API of BaseExecutor, are you back-porting this to `v2-10-test` branch (the Interface change) to know that other executor implementations need to adjust?

Otherwise, is this PR in general a bug-fix that would be valid to be ported to 2.10-line of Airflow or is this treated as new feature for Airflow 3? (I see this as 70% a bug fix for operational problem improvement and 30% a feature that myself I always found missing but never made it to the idea to contribute myself).

dstandish replied:
i am not sure what this will be backported to.  I think 2.11 is the most "by the book" but others have suggested it could or should maybe go to 2.10 @potiuk @jedcunningham

jscheffl replied:
@dstandish Triued the new cherry-picker tool and made a back-port. feel free to debate/block/approve there.

I'd like to back-port, then the deprecation is made visible early. Additionally the missing feature (in my view: bug) is resolved in 2.10.4 as well.

dstandish replied:
sounds good
